cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -0,0 +1,179 @@
+   *
+   * @param artifactId the id of the artifact to get plugins for
+   * @return an unmodifiable sorted map from plugin artifact to plugins in that artifact
+   * @throws IOException if there was an exception reading plugin metadata from the artifact store
+   */
+  public SortedMap<ArtifactDescriptor, List<PluginClass>> getPlugins(Id.Artifact artifactId) throws IOException {
+    return artifactStore.getPluginClasses(artifactId);
+  }
+
+  /**
+   * Returns a {@link Map.Entry} representing the plugin information for the plugin being requested.
+   *
+   * @param artifactId the id of the artifact to get plugins for
+   * @param pluginType plugin type name
+   * @param pluginName plugin name
+   * @param selector for selecting which plugin to use
+   * @return the entry found or {@code null} if none was found
+   * @throws IOException if there was an exception reading plugin metadata from the artifact store
+   * @throws PluginNotExistsException if no plugins of the given type and name are available to the given artifact
+   */
+  public Map.Entry<ArtifactDescriptor, PluginClass> findPlugin(Id.Artifact artifactId, final String pluginType,
+                                                               final String pluginName, PluginSelector selector)
+    throws IOException, PluginNotExistsException {
+    return selector.select(artifactStore.getPluginClasses(artifactId, pluginType, pluginName));
+  }
+
+  /**
+   * Inspects and builds plugin information for the given artifact.
+   *
+   * @param artifactId the id of the artifact to inspect and store
+   * @param artifactFile the artifact to inspect and store
+   * @param parentArtifacts artifacts the given artifact extends.
+   *                        If null, the given artifact does not extend another artifact
+   * @throws IOException if there was an exception reading from the artifact store
+   * @throws ArtifactRangeNotFoundException if none of the parent artifacts could be found
+   */
+  @VisibleForTesting
+  void inspectArtifact(Id.Artifact artifactId, File artifactFile, @Nullable Set<ArtifactRange> parentArtifacts)
+    throws IOException, ArtifactRangeNotFoundException, WriteConflictException, ArtifactAlreadyExistsException {
+
+    CloseableClassLoader parentClassLoader;
+    parentArtifacts = parentArtifacts == null ? ImmutableSet.<ArtifactRange>of() : parentArtifacts;
+    if (parentArtifacts.isEmpty()) {
+      // if this artifact doesn't extend another, use itself to create the parent classloader
+      parentClassLoader = createArtifactClassLoader(Locations.toLocation(artifactFile));
+    } else {
+      // otherwise, use any of the parent artifacts to create the parent classloader.
+      List<ArtifactDetail> parents = artifactStore.getArtifacts(parentArtifacts);
+      if (parents.isEmpty()) {
+        throw new ArtifactRangeNotFoundException(parentArtifacts);
+      }
+      parentClassLoader = createArtifactClassLoader(parents.get(0).getDescriptor().getLocation());
```

Comment:
the assumption is that any of the parents will work, otherwise it is a user error in specifying the wrong parents. If I say etl-lib extends etl-batch versions [1.0.0 , 2.0.0), any etl-batch within the range should work.